src/metatrain/utils/additive/_base_composition.py

```diff
@@ -0,0 +1,400 @@
+"""
+Contains the ``BaseCompositionModel class. This is intended for eventual porting to
+metatensor-learn. The class ``CompositionModel`` wraps this to be compatible with
+metatrain-style objects.
+"""
+from typing import Dict, List
+import torch
+
+import metatensor.torch as mts
+from metatensor.torch import TensorMap, TensorBlock, Labels, LabelsEntry
+from metatensor.torch.atomistic import System
+from metatensor.torch.learn.data import DataLoader
+
+
+class BaseCompositionModel(torch.nn.Module):
+    """Fits a composition model for a dict of targets."""
+
+    def __init__(
+        self,
+        atomic_types: List[int],
+        layouts: Dict[str, TensorMap],
+    ) -> None:
+        
+        target_names = []
+        sample_kinds = {}
+        for target_name, layout in layouts.items():  # identify target_type
+            
+            target_names.append(target_name)
+            if layout.sample_names == ["system"]:
+                sample_kinds[target_name] = "per_structure"
+            
+            elif layout.sample_names == ["system", "atom"]:
+                sample_kinds[target_name] = "per_atom"
+
+            elif layout.sample_names == [
+                "system", "first_atom", "second_atom", "cell_shift_a", "cell_shift_b", "cell_shift_c"
+            ]:
+                sample_kinds[target_name] = "per_pair"
+
+            else:
+                raise ValueError
+            
+        self.atomic_types = atomic_types
+        self.target_names = target_names
+        self.sample_kinds = sample_kinds
+
+        # find the keys that of the blocks the composition actually applies to
+        in_keys = {
+            target_name: Labels(
+                layout.keys.names,
+                torch.vstack(
+                    [
+                        key.values
+                        for key in layout.keys
+                        if _include_key(key)
+                    ]
+                )
+            )
+            for target_name, layout in layouts.items()
+        }
+
+        # filter the layout TensorMaps according to these sliced keys
+        layouts = {
+            target_name: mts.filter_blocks(layout, in_keys[target_name])
+            for target_name, layout in layouts.items()
+        }
+        self.XTX: Dict[str, TensorMap] = {
+            target_name: TensorMap(
+                layout.keys,
+                blocks=[
+                    TensorBlock(
+                        values=torch.zeros(
+                            len(self.atomic_types), len(self.atomic_types),
+                            dtype=torch.float64,
+                        ),
+                        samples=Labels(
+                            ["first_atom_type"],
```

Comment:
I'd just call this `atom_type` or `center_type`, both here and in the properties